### DISCUSSION TITLE
Thermal limiting

### DISCUSSION THREAD

**Discussion by BetterMango** [danielnilsson9/bbs-fw#45] [General]
Great project - i've been hoping for something like this to come along for years. Your efforts are much appreciated. Will donate.

Would you mind explaining how the thermal ramp down works? Does it use controller temp or motor temp? I have zero coding knowledge but I think I deciphered that a previous version cut the target current in half but not sure how it works now you've implemented gradual ramp down.

I did the shunt mod on my controller so the actual amps are twice what the controller sees so would like to attempt a build of this firmware but with a lower target current for when thermal limiting is hit. i.e if 15A is considered a good limit in my case i'd need this to be 7.5A.

**Comment by danielnilsson9** [comment]
Hi! Yes you are correct, there are some changes for next version regarding this.

From app.c:

```
// Current ramp down starts at LVC + (LVC * LVC_RAMP_DOWN_OFFSET_PERCENT / 100)
// Example:
// LVC is 42V
// 42 * 0.06 = 2.5V
// Ramp down starts at 42V + 2.5V
#define LVC_RAMP_DOWN_OFFSET_PERCENT		6	
```


So basically what you can configure using LVC_RAMP_DOWN_OFFSET_PERCENT is on what voltage current ramp down should start.
This is done in percent of LVC voltage (in percent in order to not have to worry about battery voltage, i.e. how many cells in series)

When this limit is reached current starts ramping down linearly in the interval down to a minimum percentage of maximum current which is set to 20% on line 535 (should have been a constant for that, will fix at some point). I.e. when minimum measured battery voltage is at LVC maximum output current will be at 20% of max allowed current (configured in config tool, default 30A).

The ramp is affected by the minimum measured voltage, so when high power is used and the voltage drops the minimum value will be remembered. This is done to try to avoid hitting LVC cutoff until it is completely unavoidable and battery is drained.

line 535:
`uint8_t tmp = (uint8_t)MAP32(temp_x100, (MAX_TEMPERATURE - MAX_TEMPERATURE_RAMP_DOWN_INTERVAL) * 100, MAX_TEMPERATURE * 100, 100, 20);`

See the last parameter, "20", which is 20% of maximum current -> normally 6A but 12A if you have a factor 2 on your shunt mod and have configured 30A max current.

**Reply by BetterMango** [reply]
Thanks for the explanation. I think that makes sense.
My motor has a thermistor next to the windings, connected to the hall sensor board and fed back to the controller via the JST connection. I've just noticed on your schematics you noted this as potentially a temp feed. Do you know if this is used in the original firmware?

**Reply by danielnilsson9** [reply]
Damn, I think you are right! I didn't realize there was a second thermistor, was a long time since a traced the board and I never took my motor apart to check the hall sensor board. I do remember there being a wire in the JST connector that I never figured out, but obviously I suspected it was for temperature monitor since I noted that in the drawing but never investigated further.

I guess I have some more work to do then, that is the important temperature feed to protect the windings...

**Reply by danielnilsson9** [reply]
Ehm sorry, I must have made you very confused with my answer above about low voltage limiting... which I realized is not what you asked about. But the same logic basically applies to thermal limiting too since a similar calculation is made and the minimum current when max temperature is reached is 20% of max current there too.

```
#define MAX_TEMPERATURE						75

// Current ramp down starts at MAX_TEMPERATURE - 4.
#define MAX_TEMPERATURE_RAMP_DOWN_INTERVAL	4	
```

Btw:
I have now implemented support for the temperature sensor inside the motor core, the highest temperature reading is now used for thermal limiting. Thank you for reporting this, I had totally missed it was there.

**Reply by BetterMango** [reply]
No problem - the LVC stuff was useful too and I understood the thermal limiting part which answered my question.
Glad my queries helped in a small way and thanks for implementing the winding temps so quickly. 
I'm away from my bike for a few weeks but look forward to trying this out.
